### PR TITLE
adding tensor factory for arbitrary arrays

### PR DESCRIPTION
**Pull request by MovGP0**
adding a factory method for creating a tensor from an arbitrary array type. 

*Issues*
- some data types are not supported yet, since the appropriate factory methods are missing
- method is not named `tensor`, since the compiler cannot differentiate with the `tensor(IList<> ...)` methods in some cases when given the same name (ie. when the first parameter is of type `List<T>`)
- unit tests are missing, since i was unable to compile the project on my computer (insufficient disk space for the CUDA libs)

### CODE REVIEW

**Review by NiklasGustafsson — CHANGES_REQUESTED** [review]
Looks reasonable -- I would call it 'from_array,' though.

Also, please add unit tests.

**Comment by NiklasGustafsson** [comment]
@MovGP0 -- I sent you a PR to your repo with some tests and a proposed name change.

File: src/TorchSharp/Tensor/Tensor.Factories.cs [hunk 1]
```diff
@@ -2045,6 +2045,87 @@ public static Tensor tensor(IList<System.Numerics.Complex> rawArray, long dim0,
...
+            IEnumerable<object> GetElements(Array rawArray)
+            {
+                foreach (var element in rawArray) {
+                    yield return element!;
+                }
+            }
+            var elements = GetElements(rawArray);
+
+            // call the existing factory methods to construct the tensor
+            if (t == typeof(bool)) return tensor(elements.Cast<bool>().ToArray(), shape, dtype, device, requiresGrad);
```
Comment (NiklasGustafsson) [inline src/TorchSharp/Tensor/Tensor.Factories.cs:2108]: Does the Cast<T>().ToArray() lead to a copy being made, I wonder?

File: src/TorchSharp/Tensor/Tensor.Factories.cs [hunk 2]
```diff
@@ -2045,6 +2045,87 @@ public static Tensor tensor(IList<System.Numerics.Complex> rawArray, long dim0,
...
+                    var dimLength = arr.GetLength(dim);
+                    yield return dimLength;
+                }
+            }
+            var shape = GetShape(rawArray).ToArray();
+
+            var t = rawArray.GetType().GetElementType();
+            if (t is null) throw new InvalidOperationException($"{nameof(rawArray)}.GetType().GetElementType() returned null.");
+
+            IEnumerable<object> GetElements(Array rawArray)
```
Comment (NiklasGustafsson) [inline src/TorchSharp/Tensor/Tensor.Factories.cs:2099]: This function seem unnecessary. 'Array' already implements IEnumerable, so this just adds overhead, I think.

File: src/TorchSharp/Tensor/Tensor.Factories.cs [hunk 3]
```diff
@@ -2045,6 +2045,87 @@ public static Tensor tensor(IList<System.Numerics.Complex> rawArray, long dim0,
...
+            if (t == typeof(long)) return tensor(elements.Cast<long>().ToArray(), shape, dtype, device, requiresGrad);
+#if NET50_OR_GREATER
+            // TODO: implement the required factory method
+            // if (t == typeof(half)) return tensor(elements.Cast<half>().ToArray(), shape, dtype, device, requiresGrad);
+#endif
+            if (t == typeof(float)) return tensor(elements.Cast<float>().ToArray(), shape, dtype, device, requiresGrad);
+            if (t == typeof(double)) return tensor(elements.Cast<double>().ToArray(), shape, dtype, device, requiresGrad);
+            // TODO: implement the required factory methods
+            // if (t == typeof(Tuple<float, float>)) return tensor(elements.Cast<Tuple<float, float>>().ToArray(), shape, dtype, device, requiresGrad);
+            // if (t == typeof(Tuple<double, double>)) return tensor(elements.Cast<Tuple<double, double>>().ToArray(), shape, dtype, device, requiresGrad);
```
Comment (NiklasGustafsson) [inline src/TorchSharp/Tensor/Tensor.Factories.cs:2122]: We don't use (double,double) to represent 64-bit complex, and the (float,float) factory methods should already exist.

**Comment by NiklasGustafsson** [comment]
@MovGP0 -- just a friendly "ping." I think this is a great addition to the APIs, but I had some comments on it last week.

**Comment by MovGP0** [comment]
> @MovGP0 -- just a friendly "ping." I think this is a great addition to the APIs, but I had some comments on it last week.

much appreciated. Just very busy these weeks.

File: src/TorchSharp/Tensor/Tensor.Factories.cs (hunk 1 above)
Comment (MovGP0) [reply src/TorchSharp/Tensor/Tensor.Factories.cs:2108]: yes, the `.ToArray()` always creates a new array: 
 ```csharp
 var array = new int[]{1,2,3,4};
var newArray = array.Cast<int>().ToArray();
Console.WriteLine(ReferenceEquals(array, newArray)); // false
```

 ```csharp
 var array = new int[]{1,2,3,4};
var newArray = array.ToArray();
Console.WriteLine(ReferenceEquals(array, newArray)); // false
```

**Comment by MovGP0** [comment]
> Looks reasonable -- I would call it 'from_array,' though.

I was thinking about that and opted against it. The more verbose `tensor_from_array` makes the method easier to discover (think of the autocomplete when someone types `torch.tens...`) and makes its function more clear.

**Comment by NiklasGustafsson** [comment]
> * fixing documentation to reflect the name change to `from_array`

I thought you didn't agree with that.

**Comment by NiklasGustafsson** [comment]
I'm going to merge this one as soon as the restarted checks pass. We can always change names later, before the next release.

**Review by NiklasGustafsson — APPROVED** [review]

**Pull request merged**